internal/jobworker/job/job.go

```diff
@@ -0,0 +1,302 @@
+	// ID is a unique identifier.
+	ID uuid.UUID
+	// Owner is the user responsible for Job instance creation.
+	Owner string
+
+	cmd      reexec.Command
+	status   Status
+	exitCode int
+
+	// context.Context is usually utilized at the function level. However, here
+	// it is being used to coordinate the cancelling of all async Job resources.
+	ctx    context.Context
+	cancel context.CancelFunc
+
+	exec                    *exec.Cmd
+	cmdIn, cmdOut           io.WriteCloser
+	continueIn, continueOut io.WriteCloser
+}
+
+// StreamOutput streams Job's output to the passed stream channel in chunks of
+// size chunkSize. StreamOutput will return if either of the following
+// circumstances occur:
+//
+// 1) The ctx is cancelled.
+// 2) The Job is no longer running and the end of the output is reached.
+func (j Job) StreamOutput(ctx context.Context, stream chan<- []byte, chunkSize int) error {
+	ctx, cancel := context.WithCancel(ctx)
+	defer cancel()
+
+	fd, err := os.Open(output.File(j.ID))
+	if err != nil {
+		return fmt.Errorf("open job output; error: %w", err)
+	}
+	go func() {
+		<-ctx.Done()
+		fd.Close()
+	}()
+
+	b := make([]byte, chunkSize)
+	for {
+		n, err := fd.Read(b)
+		// If any bytes were read at all, write to stream.
+		if n > 0 {
+			select {
+			case <-ctx.Done():
+				return ctx.Err()
+			case stream <- b[:n]:
+			}
+		}
+		// If context has been cancelled return to caller.
+		if errors.Is(ctx.Err(), context.Canceled) {
+			return ctx.Err()
+		}
+		// If EOF and job is running, wait for output from job.
+		if errors.Is(err, io.EOF) && j.Status() == Running {
+			// TODO: This is a placeholder. Streaming implementation using inotify
+			// API will be in another PR. This will be replaced by something like
+			// select {
+			// case <-ctx.Done():
+			//   return ctx.Err()
+			// case <-j.watchOutput()
+			// }
+			time.Sleep(time.Second)
+			continue
+		}
+		/// If EOF and job is not running, return.
+		if errors.Is(err, io.EOF) {
+			return nil
+		}
+		if err != nil {
+			return fmt.Errorf("read job output; error: %w", err)
+		}
+	}
+}
+
+// Status retrieves the Job status.
+func (j Job) Status() Status {
+	j.mutex.RLock()
+	defer j.mutex.RUnlock()
+	return j.status
+}
+
+// ExitCode retrieves the Job exit code.
+func (j Job) ExitCode() int {
+	j.mutex.RLock()
+	defer j.mutex.RUnlock()
+	return j.exitCode
+}
+
+// cleanup releases all resources tied to the Job. cleanup should be called
+// once the Job is no longer being used.
+func (j Job) cleanup() {
+	j.stop()
+
+	closers := []io.Closer{
+		j.cmdIn,
+		j.cmdOut,
+		j.continueIn,
+		j.continueOut,
+	}
+
+	for _, closer := range closers {
+		closer.Close()
+	}
+}
+
+// start launches the Job.
+func (j *Job) start() error {
+	logger.Infof("starting Job; ID: %v", j.ID)
+
+	if err := j.exec.Start(); err != nil {
+		return fmt.Errorf("start child process; error: %w", err)
+	}
+
+	// Write job details to cmdIn pipe. Child process will read and launch
+	// grandchild process.
+	go func() {
```

Comment:
I think we could do away with this goroutine altogether. Delays start() a bit, but I think I'd rather take the delay.